### PR TITLE
`quilt export` original filepaths, additional testing, and minor updates

### PR DESCRIPTION
**Pull request by eode**
* Export now uses exact filenames from original build
* Handle export when a non-GroupNode is specified (single-file export)
* Notify user when result includes no exportable files
* Test subpkg export, single-file export, filters, and mappers

Thanks to to akarve for letting me know that the original filenames were actually retained, and where.

### CODE REVIEW

**Comment by eode** [comment]
Fixing test failure...

**Comment by eode** [comment]
There are a lot of little test failures I need to resolve -- some simply caught by travis, others I've introduced, and others on the tensorflow branch that I've merged in.  Some are a little off-topic for this PR, but it's all going into tensorflow before merge into master, so I'm including them if that's ok.

**Comment by eode** [comment]
Got hung up on a weird test fail issue that was just due to problematic / conflicting data in local storage.

Needed higher python version compat install.  Fixed.
Cleaned up compat module.

File: compiler/quilt/test/test_build.py [hunk 1]
```diff
@@ -121,7 +121,7 @@ def test_build_group_args(self):
         build.build_package('groups', 'pkg', path)
 
         from quilt.data.groups import pkg
-        
+
```
Comment (eode) [inline compiler/quilt/test/test_build.py]: Apologies re: whitespace fix in commit with README_md issue.

**Comment by eode** [comment]
@akarve @kevinemoore 

This is ready for review.

**Comment by eode** [comment]
Had a pretty nasty time with some inconsistent retention of data -- ended up just being `force=True` on `quilt.install()` in train.py -- which automatically overwrites any local changes with the remote repository state.

Also, fixed an area where _set() isn't adding metadata as it should.  Checking with Aneesh and Dima regarding this.

**Comment by eode** [comment]
Knocked a couple TODOs off that were quick and were issues we'd run into a couple times -- specifically, raising an exception in _set() when an unknown type is given, and checking node names more completely.

In that, I've added a `tools.utils.is_nodename(string)` function that returns True if the given stringlike object is a valid Nodename (is a valid python identifier, and doesn't start with an underscore).  

Python 2.7's str/unicode classes don't have an 'isidentifier' function, so I also added an `is_identifier(s)` function in utils to check if a stringlike is a valid python identifier.

**Comment by eode** [comment]
Failed on travis for 3.4 due to a timed-out pandas install, but a re-run went fine (although the re-run didn't show up here).

**Comment by eode** [comment]
Also in this PR, metadata and metadata['q_path'] are now required for FileNode objects -- checked with Aneesh.

**Comment by eode** [comment]
This is ready for review again.

File: compiler/quilt/nodes.py [hunk 2]
```diff
@@ -131,19 +131,26 @@ def _set(self, path, value):
         assert isinstance(path, list) and len(path) > 0
 
         if isinstance(value, pd.DataFrame):
+            # TODO: what about dataframe metadata?
```
Comment (asah) [inline compiler/quilt/nodes.py]: raise NotImplemented for this initial commit.

File: compiler/quilt/nodes.py [hunk 3]
```diff
@@ -131,19 +131,26 @@ def _set(self, path, value):
...
             core_node = core.TableNode(hashes=[])
         elif isinstance(value, string_types):
-            core_node = core.FileNode(hashes=[])
+            # If metadata isn't given, the FileNode won't retain the file path.
+            metadata = {'q_path': value, 'q_target': 'file', 'q_ext': ''}
+            core_node = core.FileNode(hashes=[], metadata=metadata)
         elif isinstance(value, bytes):
+            # TODO: Are these bytes a filename?  If this is this re: 2 vs 6, it's already handled via string_types above
```
Comment (asah) [inline compiler/quilt/nodes.py]: pls ask @dimaryaz or @kevinemoore via Slack

File: compiler/quilt/test/test_build.py [hunk 4]
```diff
@@ -166,7 +166,8 @@ def test_generate_buildfile(self):
         assert os.path.exists(buildfilepath)
         build.build_package('test_generated', 'generated', buildfilepath)
         os.remove(buildfilepath)
-        from quilt.data.test_generated.generated import bad, foo, nuts, README
+        # XXX: building now produces README_md instead of README -- intentional?
+        from quilt.data.test_generated.generated import bad, foo, nuts, README_md
```
[the remark this answers is not in the document]
Comment (asah) [reply compiler/quilt/test/test_build.py]: y, intentional

File: compiler/quilt/test/test_command.py [hunk 5]
```diff
@@ -523,41 +523,104 @@ def test_rm_doesnt_break_cache(self):
...
             ]
 
+        test_data = [
+            (single_name, single_bytes),
+            ('readme.md', os.urandom(200)),
+            # these are invalid python identifiers, but should be handled without issue
+            ('3-bad-identifier/bad_parent_identifier.html', os.urandom(100)),
+            ('3-bad-identifier/9{}bad-identifier.html', os.urandom(100)),
+            ] + subdir_test_data
+
```
Comment (asah) [inline compiler/quilt/test/test_command.py]: also add test cases for:
- overwriting existing dirs/files
- improper permissions to create the subdir
- (if easy) full disk

File: examples/tensorflow/tensorflow-cifar-10/train-short.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
+# This should just be 'train.py' with a shorter interval.  Feel free to delete or modify.
```
Comment (asah) [inline examples/tensorflow/tensorflow-cifar-10/train-short.py]: pls remove from PR ?

File: examples/tensorflow/tensorflow-cifar-10/train.py [hunk 7]
```diff
@@ -13,7 +13,8 @@
 QUILT_PKG = 'asah/cifar10'
 
 print("Connecting to Quilt...")
-quilt.install(QUILT_PKG, force=True)
+# force=False because otherwise local data in storage is overwritten
+quilt.install(QUILT_PKG, force=False)
```
Comment (asah) [inline examples/tensorflow/tensorflow-cifar-10/train.py]: actually, force=True avoids the interactive display, which is why I did it

**Pull request merged**